bank/handlers.go

```diff
@@ -0,0 +1,380 @@
+		}
+		json.NewEncoder(w).Encode(result)
+	case http.MethodPost:
+		err := us.InsertAccount(request)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+		}
+	case http.MethodPatch:
+		switch request.TransactionType {
+		case "deposit":
+			result, err := us.GetAccounts(request)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusBadRequest)
+			}
+
+			for _, current := range result {
+				if current.Currency == request.Currency {
+					request.Amount = current.Amount + request.Amount
+					err := us.UpdateAccount(request)
+					if err != nil {
+						http.Error(w, err.Error(), http.StatusBadRequest)
+					}
+				}
+			}
+		case "withdraw":
+			result, err := us.GetAccounts(request)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusBadRequest)
+			}
+
+			for _, current := range result {
+				if current.Currency == request.Currency {
+					if current.Amount >= request.Amount {
+						request.Amount = current.Amount - request.Amount
+						err := us.UpdateAccount(request)
+						if err != nil {
+							http.Error(w, err.Error(), http.StatusBadRequest)
+						}
+					} else {
+						http.Error(w, "cannot withdraw more than have", http.StatusBadRequest)
+					}
+				}
+			}
+		}
+	case http.MethodDelete:
+		err := us.DeleteAccount(request)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+		}
+	}
+
+	err := us.UpdateSession(request)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+
+	us.renewSessionCounter <- true
+
+	gorillaSession.Options.MaxAge = 300
+	gorillaSession.Options.Path = "/"
+
+	gorillaSession.Save(r, w)
+}
+
+func (us *UserRepo) SignInHandler(w http.ResponseWriter, r *http.Request) {
+	u := context.Get(r, "user")
+	gsr := context.Get(r, "gsr")
+	s := context.Get(r, "session")
+
+	user := u.(User)
+	gorSessRepo := gsr.([]Session)
+	gorillaSession := s.(*sessions.Session)
+
+	if len(gorSessRepo) != 0 && string(gorSessRepo[0].ID) == gorillaSession.Values[user.Username] && !time.Now().Local().After(gorSessRepo[0].Expires) {
+		http.Error(w, "session not expired", 400)
+		return
+	}
+
+	err := us.authenticate(user)
+	if err != nil {
+		http.Error(w, err.Error(), 404)
+		return
+	}
+
+	if len(gorillaSession.ID) == 0 {
+		gorillaSession.Options.MaxAge = 300
+		gorillaSession.Options.Path = "/"
+
+		err := us.InsertSession(user, &gorSessRepo)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+	}
+
+	gorillaSession.Values[user.Username] = string(gorSessRepo[0].ID)
+	gorillaSession.Save(r, w)
+
+	go us.voidSessionAfter(user, w)
+
+	w.WriteHeader(http.StatusOK)
+	b := []byte(user.Username)
+	w.Write(b)
+}
+
+func (us *UserRepo) SignUpHandler(w http.ResponseWriter, r *http.Request) {
+	var u User
+
+	json.NewDecoder(r.Body).Decode(&u)
+
+	err := u.hashGenerator()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+
+	err = us.InsertUser(u)
```

Comment:
It's better to perform check queries instead of verifying of errors that are returned from the datastore during the insert.